{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm",
      "gpuType": "T4"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Structure and complex prediction using Boltz\n",
        "\n",
        "One of the challenges in structure prediction is structures of biomolecules interacting with other molecules, such as a enzyme bound to a substrate or a protein-DNA complex. With new methods such as [AlphaFold3](https://alphafoldserver.com/welcome) and the ever expanding structural data in the [Protein Data Bank](https://www.rcsb.org/), it is possible to (cautiously) model these complexes. Additionally, resources such as Google Colab (you are in one!) allow for greater access to these methods in scientific study.\n",
        "\n",
        "**[Boltz-1](https://github.com/jwohlwend/boltz)** is a modeling algorithem built on AlphaFold3 that results in similar prediction quality. So why use it? The current iteration of the AlphaFold3 server (as of 2025) is highly restrained in the modeling options such as post-translational modifications and ligands. Boltz-1 has fewer restrictions on the options allowing for greater opportunity to model complexes.\n",
        "This colab runs a version of Boltz-1 and reduces the input complexity to allow modeling of protein-ligand complexes. It is an adaptation of the notebook generated by the [Colabfold](https://github.com/sokrypton/ColabFold) group to make is accessible in the biochemistry classroom.\n",
        "\n",
        "##**Please make a copy of this colab for your personal use!!**\n",
        "\n",
        "A few things to start:\n",
        "\n",
        "1.   These lessons only work in Google Chrome\n",
        "2.   If you want to save your progress, go to File> Save a Copy in Drive; then locate a spot in your Drive folder\n",
        "3.   Clicking the \"play\" button to the top left of a code block runs the code. Sometimes you can see the code and interact with it. However, if the code is hidden  it is to run adminstrative tasks in the background and you do not need to worry unless you are interested.\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "**A few notes:**\n",
        "- Boltz-1 does not minimize structures after modeling and is known to produce overlapping chains or steric clashes in amino acids. If this is an issue, try modeling through another method or run an [MD simulation](https://colab.research.google.com/drive/1-Nr3oze1IWND4-eTci9TyykFhtx9fVuX?usp=sharing) to rectify issues.\n",
        "\n",
        "- Long proteins or large complexes (~>2000 amino acids in total) may exceed the memory limits of the colab.\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "**Acknowledgements**\n",
        "\n",
        "This Colab was adapted by Chris Berndsen at James Madison University (ver 2 2025) from a notebook written by [Colabfold](https://github.com/sokrypton/ColabFold) in 2024. The [Colabfold version of Boltz-1](https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/Boltz1.ipynb) offers more options than presented here.\n"
      ],
      "metadata": {
        "id": "KNOrLaJFdiA-"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Getting started\n",
        "\n",
        "To start modeling in Boltz, you need the sequence of your protein(s) in one letter amino acid code. If you are predicting a ligand bound complex you will also need the SMILES string(s), likely from [PubChem](https://pubchem.ncbi.nlm.nih.gov/).\n",
        "\n",
        "\n",
        "\n",
        "1.   Paste your first protein sequence into the `query sequence` box\n",
        "2.   If you have a multimeric protein, type a `:` after the first sequence and then paste in the next sequence. Repeat this until all protein sequences are in.\n",
        "**NOTE:** You must have a sequence for each member of the complex, even if the protein is composed of multiple copies of the same protein.\n",
        "\n",
        "3.   Add any ligands. If you are not predicting ligands, make sure the `ligand_input` section is completely blank except for the Insert text here.\n",
        "4.   Press the play button. This step should take less than 10 seconds to run.\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "cYrtgbtIDgX6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Input protein sequence(s)\n",
        "from google.colab import files\n",
        "import os\n",
        "import re\n",
        "import hashlib\n",
        "import random\n",
        "import requests\n",
        "from string import ascii_uppercase\n",
        "\n",
        "# Function to add a hash to the jobname\n",
        "def add_hash(x, y):\n",
        "    return x + \"_\" + hashlib.sha1(y.encode()).hexdigest()[:5]\n",
        "\n",
        "# User inputs\n",
        "query_sequence = ''  #@param {type:\"string\"}\n",
        "#@markdown  - Use `:` to specify inter-protein chainbreaks for **modeling complexes** (supports homo- and hetero-oligomers). For example **PI...SK:PI...SK** for a homodimer\n",
        "ligand_input = ''  #@param {type:\"string\"}\n",
        "#@markdown  - Use `:` to specify multiple ligands as smile strings\n",
        "##ligand_input_ccd = 'SAH'  #@param {type:\"string\"}\n",
        "##@markdown - Use `:` to specify multiple ligands as CCD codes (three-letter codes)\n",
        "##ligand_input_common_name = ''  #@param {type:\"string\"}\n",
        "##@markdown - Use `:` to specify multiple ligands with their common name (e.g. Aspirin; SMILES fetched from [PubChem](https://pubchem.ncbi.nlm.nih.gov) API)\n",
        "#dna_input = ''  #@param {type:\"string\"}\n",
        "##@markdown - Use `:` to specify multiple DNA sequences\n",
        "jobname = ''  #@param {type:\"string\"}\n",
        "\n",
        "# Clean up the query sequence and jobname\n",
        "query_sequence = \"\".join(query_sequence.split())\n",
        "ligand_input = \"\".join(ligand_input.split())\n",
        "#ligand_input_ccd = \"\".join(ligand_input_ccd.split())\n",
        "#ligand_input_common_name = \"\".join(ligand_input_common_name.split())\n",
        "#dna_input = \"\".join(dna_input.split())\n",
        "basejobname = \"\".join(jobname.split())\n",
        "basejobname = re.sub(r'\\W+', '', basejobname)\n",
        "jobname = add_hash(basejobname, query_sequence)\n",
        "\n",
        "# Check if a directory with jobname exists\n",
        "def check(folder):\n",
        "    return not os.path.exists(folder)\n",
        "\n",
        "if not check(jobname):\n",
        "    n = 0\n",
        "    while not check(f\"{jobname}_{n}\"):\n",
        "        n += 1\n",
        "    jobname = f\"{jobname}_{n}\"\n",
        "\n",
        "# Make directory to save results\n",
        "os.makedirs(jobname, exist_ok=True)\n",
        "\n",
        "from string import ascii_uppercase\n",
        "\n",
        "# Split sequences on chain breaks\n",
        "protein_sequences = query_sequence.strip().split(':') if query_sequence.strip() else []\n",
        "ligand_sequences = ligand_input.strip().split(':') if ligand_input.strip() else []\n",
        "#ligand_sequences_ccd = ligand_input_ccd.strip().split(':') if ligand_input_ccd.strip() else []\n",
        "#ligand_sequences_common_name = ligand_input_common_name.strip().split(':') if ligand_input_common_name.strip() else []\n",
        "#dna_sequences = dna_input.strip().split(':') if dna_input.strip() else []\n",
        "\n",
        "def get_smiles(compound_name):\n",
        "    autocomplete_url = f\"https://pubchem.ncbi.nlm.nih.gov/rest/autocomplete/compound/{compound_name}/json?limit=1\"\n",
        "    autocomplete_response = requests.get(autocomplete_url)\n",
        "    if autocomplete_response.status_code != 200:\n",
        "        return None\n",
        "\n",
        "    autocomplete_data = autocomplete_response.json()\n",
        "    if autocomplete_data.get(\"status\", {}).get(\"code\") != 0 or autocomplete_data.get(\"total\", 0) == 0:\n",
        "        return None\n",
        "\n",
        "    suggested_compound = autocomplete_data.get(\"dictionary_terms\", {}).get(\"compound\", [])\n",
        "    if not suggested_compound:\n",
        "        return None\n",
        "    suggested_compound_name = suggested_compound[0]\n",
        "\n",
        "    smiles_url = f\"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/name/{suggested_compound_name}/property/CanonicalSMILES/JSON\"\n",
        "    smiles_response = requests.get(smiles_url)\n",
        "    if smiles_response.status_code != 200:\n",
        "        return None\n",
        "\n",
        "    smiles_data = smiles_response.json()\n",
        "    properties = smiles_data.get(\"PropertyTable\", {}).get(\"Properties\", [])\n",
        "    if len(properties) == 0:\n",
        "        return None\n",
        "\n",
        "    return properties[0].get(\"CanonicalSMILES\")\n",
        "\n",
        "# Initialize chain labels starting from 'A'\n",
        "chain_labels = iter(ascii_uppercase)\n",
        "\n",
        "fasta_entries = []\n",
        "csv_entries = []\n",
        "chain_label_to_seq_id = {}\n",
        "seq_to_seq_id = {}\n",
        "seq_id_counter = 0  # Counter for unique sequences\n",
        "\n",
        "# Process protein sequences\n",
        "for seq in protein_sequences:\n",
        "    seq = seq.strip()\n",
        "    if not seq:\n",
        "        continue  # Skip empty sequences\n",
        "    chain_label = next(chain_labels)\n",
        "    # Check if sequence has been seen before\n",
        "    if seq in seq_to_seq_id:\n",
        "        seq_id = seq_to_seq_id[seq]\n",
        "    else:\n",
        "        seq_id = f\"{jobname}_{seq_id_counter}\"\n",
        "        seq_to_seq_id[seq] = seq_id\n",
        "        seq_id_counter += 1\n",
        "        # For CSV file (for ColabFold), add only unique sequences\n",
        "        csv_entries.append((seq_id, seq))\n",
        "    chain_label_to_seq_id[chain_label] = seq_id\n",
        "    # For FASTA file\n",
        "    msa_path = os.path.join(jobname, f\"{seq_id}.a3m\")\n",
        "    header = f\">{chain_label}|protein|{msa_path}\"\n",
        "    sequence = seq\n",
        "    fasta_entries.append((header, sequence))\n",
        "\n",
        "# Process ligand sequences (assumed to be SMILES strings)\n",
        "for lig in ligand_sequences:\n",
        "    lig = lig.strip()\n",
        "    if not lig:\n",
        "        continue  # Skip empty ligands\n",
        "    chain_label = next(chain_labels)\n",
        "    lig_type = 'smiles'\n",
        "    header = f\">{chain_label}|{lig_type}\"\n",
        "    sequence = lig\n",
        "    fasta_entries.append((header, sequence))\n",
        "\n",
        "# Write the CSV file for ColabFold\n",
        "queries_path = os.path.join(jobname, f\"{jobname}.csv\")\n",
        "with open(queries_path, \"w\") as text_file:\n",
        "    text_file.write(\"id,sequence\\n\")\n",
        "    for seq_id, seq in csv_entries:\n",
        "        text_file.write(f\"{seq_id},{seq}\\n\")\n",
        "\n",
        "# Write the FASTA file\n",
        "queries_fasta = os.path.join(jobname, f\"{jobname}.fasta\")\n",
        "with open(queries_fasta, 'w') as f:\n",
        "    for header, sequence in fasta_entries:\n",
        "        f.write(f\"{header}\\n{sequence}\\n\")\n",
        "\n",
        "# Optionally, print the output for verification\n",
        "#print(f\"Generated FASTA file '{queries_fasta}':\\n\")\n",
        "#for header, sequence in fasta_entries:\n",
        "#    print(f\"{header}\\n{sequence}\\n\")\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "AcYvVeDESi2a"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4eXNO1JJHYrB",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Install dependencies\n",
        "#@markdown This next step imports and installs the code necessary to run the prediction. No user inputs are needed and it should run in less than 5 minutes.\n",
        "%%time\n",
        "import os\n",
        "if not os.path.isfile(\"COLABFOLD_READY\"):\n",
        "  print(\"installing colabfold...\")\n",
        "  os.system(\"pip install -q --no-warn-conflicts 'colabfold[alphafold-minus-jax] @ git+https://github.com/sokrypton/ColabFold'\")\n",
        "  if os.environ.get('TPU_NAME', False) != False:\n",
        "    os.system(\"pip uninstall -y jax jaxlib\")\n",
        "    os.system(\"pip install --no-warn-conflicts --upgrade dm-haiku==0.0.10 'jax[cuda12_pip]'==0.3.25 -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html\")\n",
        "  else:\n",
        "    os.system(\"pip install --no-warn-conflicts --upgrade dm-haiku==0.0.13\")\n",
        "  os.system(\"ln -s /usr/local/lib/python3.*/dist-packages/colabfold colabfold\")\n",
        "  os.system(\"ln -s /usr/local/lib/python3.*/dist-packages/alphafold alphafold\")\n",
        "  os.system(\"touch COLABFOLD_READY\")\n",
        "\n",
        "if not os.path.isfile(\"BOLZ_READY\"):\n",
        "  os.system(\"apt-get install -y aria2\")\n",
        "  os.system(\"pip install -q --no-warn-conflicts boltz\")\n",
        "  os.system(\"mkdir weights\")\n",
        "  os.system(\"aria2c -d weights -x8 -s8 https://colabfold.steineggerlab.workers.dev/boltz1.ckpt\")\n",
        "  os.system(\"aria2c -d weights -x8 -s8 https://colabfold.steineggerlab.workers.dev/ccd.pkl\")\n",
        "  os.system(\"touch BOLZ_READY\")"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Multiple sequence alignment\n",
        "One thing prediction methods use is an MSA or multiple sequence alignment to identify sequence conservation and co-evolution. This latter event comes from the idea that sequences that interact with each other co-evolve or change in complimentary ways to maintain the interaction site. Many sequences are needed and Boltz-1 uses the [mmseqs](https://github.com/soedinglab/MMseqs2) method to match and align many sequences. This method is similar to [NCBI BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi) but runs faster. Additionally, for predictions of complexes that are not homomers (meaning not composed of many of the same protein), the sequences have to be matched by taxonomy so that the co-evolution can be followed accurately.\n",
        "\n",
        "This step runs in less than 5 minutes usually."
      ],
      "metadata": {
        "id": "_NY6M0ztIaKA"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Generate MSA with ColabFold\n",
        "!colabfold_batch \"{queries_path}\" \"{jobname}\" --msa-only"
      ],
      "metadata": {
        "id": "4aFDR4IhRe6y",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Predicting using Boltz-1\n",
        "\n",
        "We are now ready to run Boltz-1 prediction. The algorithm will take the input information and the MSA and create a model of the biomolecule(s). While too complex to describe in this short box, effectively the workflow is to represent amino acids/ligands as tokens and progressively move these tokens in 3-D space until they result in a model which is reasonable from a protein structure standpoint. The system was trained on the information in the PDB and thus the constraints are based in reality. Thus, sequences and complexes for which there is existing or similar data will be predicted with higher confidence. Linked is a video which explores the [background on Boltz-1](https://youtu.be/mN5pXxRW4bA?si=4x4XtGIPTN469x6c) in more detail.\n",
        "\n",
        "The time that this step takes varies a lot. In an example run, a homodimer of ~650 amino acids plus two ligands took about 15 minutes to run."
      ],
      "metadata": {
        "id": "-iEx2vZlKBWt"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Predict structure using boltz\n",
        "!boltz predict --out_dir \"{jobname}\" \"{jobname}/{jobname}.fasta\" --cache weights"
      ],
      "metadata": {
        "id": "bgaBXxXtIAu9",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Visualizing the result\n",
        "The next block will show the resulting model in 4 different ways:\n",
        "\n",
        "\n",
        "*   Top left is the model colored by confidence.\n",
        "*   Top right is the model in surface representation colored by confidence\n",
        "*  Bottom left is model colored in rainbow in each\n",
        "*  Bottom right is the model showing ligands as spheres\n",
        "\n",
        "\n",
        "You can drag and rotate these structures and Zoom in on specific features if you wish."
      ],
      "metadata": {
        "id": "FtlJaNbfRIyE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Display 3D structure {run: \"auto\"}\n",
        "import py3Dmol\n",
        "import glob\n",
        "import matplotlib.pyplot as plt\n",
        "from colabfold.colabfold import plot_plddt_legend\n",
        "from colabfold.colabfold import pymol_color_list, alphabet_list\n",
        "color = \"lDDT\"\n",
        "show_sidechains = False\n",
        "\n",
        "\n",
        "pdb_filename = os.path.join(jobname, f'boltz_results_{jobname}', 'predictions', jobname, '*.cif')\n",
        "pdb_file = glob.glob(pdb_filename)\n",
        "\n",
        "plot_plddt_legend().show()\n",
        "model_name = jobname\n",
        "view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js',viewergrid= (2,2))\n",
        "view.addModel(open(pdb_file[0],'r').read(),'cif')\n",
        "view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'ROYGB','min':50,'max':90}}}, viewer = (0,0))\n",
        "view.setStyle({'cartoon': {'color':'spectrum'}}, viewer = (1,0))\n",
        "view.setStyle({'cartoon': {'color':'white', 'width':'0.5'}}, viewer = (0,1))\n",
        "BB = ['C', 'O', 'N']\n",
        "view.addStyle({'and':[{'resn':[\"GLY\",\"PRO\"],'invert':True},{'atom':BB,'invert':True}]},\n",
        "                   {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}}, viewer= (0,1))\n",
        "view.addSurface(py3Dmol.VDW,{'opacity':0.6,'colorscheme': {'prop':'b','gradient': 'ROYGB','min':50,'max':90}}, viewer = (0,1))\n",
        "view.setStyle({'cartoon': {'color':'grey'}}, viewer = (1,1))\n",
        "view.setStyle({'resn':'LIG'}, {'sphere':{'colorscheme':'coralCarbon'}}, viewer = (1,1))\n",
        "\n",
        "view.zoomTo()\n",
        "view.render()\n"
      ],
      "metadata": {
        "id": "3bHTRIwzaezg",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "## Modeling Statistics\n",
        "\n",
        "In addition the model, we also can see the statistics of the modeling. These scores are confidence scores in the structure, and if your modeled multiple molecules, the complex. These scores run from 0 to 1 with 0 meaning no confidence and 1 meaning perfect confidence.\n",
        "\n",
        "This [site](https://neurosnap.ai/blog/post/interpreting-boltz-1-alphafold3-metrics-and-visualizations-on-neurosnap/675b7b92375d5ec1fde492ef) has more information on some of the statistics\n",
        "\n",
        "\n",
        "\n",
        "### Scores\n",
        "\n",
        "The main scores to look at when assessing the modeling are:\n",
        "\n",
        "1.   Confidence score: Similar to the pLDDT score in AlphaFold3\n",
        "2.   ipTM: interface predicted template modelling\n",
        "3.   ligand_ipTM: interface predicted template modelling for ligand\n",
        "4.   complex_pde: Predicted Docking Error describes uncertainty in 3-d arrangement of chains\n",
        "\n",
        "\n",
        "\n",
        "### Are the data trustworthy?\n",
        "\n",
        "*   Confidence score > 0.9: Very confident\n",
        "*   0.9 > Confidence score > 0.8: Confident\n",
        "*   0.8 > Confidence score > 0.6: Cautiously confident\n",
        "*   0.6 > Confidence score > 0.5: Skeptical\n",
        "*   Confidence score < 0.5: Try another method\n",
        "\n",
        "For a biomolecule complex:\n",
        "\n",
        "*  ipTM > 0.5: Likely on right track for complex\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "***Remember***, these models are trained on experimental structures, so if you are working with a sequence with similarity to known structures, you are almost guaranteed to get a decent score. This can complicate studies on mutants as the algorithm will always lean toward an existing structure over adapting to the sequence change. **Regardless of the scores, these are estimates and should be supported with experimental data whenever possible.**\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "WXNLAFZJjQig"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Modeling statistics\n",
        "#import pandas as pd\n",
        "#json = glob.glob(os.path.join(jobname, f'boltz_results_{jobname}', 'predictions', jobname, '*.json'))\n",
        "\n",
        "#jsondf = pd.read_json(json[0])\n",
        "\n",
        "#jsondf.to_csv(os.path.join(jobname, f'boltz_results_{jobname}', 'predictions', jobname, 'stats.csv'))\n",
        "\n",
        "# remove the zeros from jsondf\n",
        "#jsondf = jsondf.T\n",
        "#jsondf = jsondf[jsondf[0] != 0]\n",
        "\n",
        "#print(jsondf.to_string())\n",
        "\n",
        "@title Modeling statistics\n",
        "import json\n",
        "import csv\n",
        "import glob\n",
        "import os\n",
        "\n",
        "def json_to_csv(json_file_path, csv_file_path):\n",
        "    \"\"\"\n",
        "    Converts a JSON file to a CSV file.\n",
        "\n",
        "    Args:\n",
        "        json_file_path (str): The path to the JSON file.\n",
        "        csv_file_path (str): The path to the CSV file to be created.\n",
        "    \"\"\"\n",
        "    with open(json_file_path, 'r') as json_file:\n",
        "        data = json.load(json_file)\n",
        "\n",
        "    if not data:\n",
        "        print(\"JSON file is empty.\")\n",
        "        return\n",
        "\n",
        "    if isinstance(data, dict):\n",
        "        data = [data]\n",
        "\n",
        "    with open(csv_file_path, 'w', newline='') as csv_file:\n",
        "        writer = csv.writer(csv_file)\n",
        "\n",
        "        header = list(data[0].keys()) if data else []\n",
        "        writer.writerow(header)\n",
        "\n",
        "        for row in data:\n",
        "            writer.writerow(row.values())\n",
        "\n",
        "\n",
        "# Example usage:\n",
        "# Get a list of JSON files using glob\n",
        "json_files = glob.glob(os.path.join(jobname, f'boltz_results_{jobname}', 'predictions', jobname, '*.json'))\n",
        "\n",
        "# Iterate through the list and process each file\n",
        "for json_file in json_files:  # Added loop to handle multiple JSON files\n",
        "    csv_file = os.path.join(jobname, f'boltz_results_{jobname}', 'predictions', jobname, 'stats.csv') # Output CSV file\n",
        "    json_to_csv(json_file, csv_file)  # Pass the individual file path\n",
        "\n",
        "print(\"Conversion completed.\")\n",
        "\n",
        "def print_csv_table(file_path):\n",
        "    \"\"\"Prints a CSV file as a formatted table.\"\"\"\n",
        "    with open(file_path, 'r') as file:\n",
        "        reader = csv.reader(file)\n",
        "\n",
        "        # Get the header row and calculate column widths\n",
        "        header = next(reader)\n",
        "        column_widths = [max(len(str(item)) for item in col) for col in zip(*([header] + list(reader)))]\n",
        "\n",
        "        # Print the header\n",
        "        for i, col in enumerate(header):\n",
        "            print(f\"{col:<{column_widths[i]}}\", end=\"  \")\n",
        "        print()\n",
        "        print(\"-\" * (sum(column_widths) + 4 * len(header))) # Print a separator line\n",
        "\n",
        "        # Print the data rows\n",
        "        file.seek(0) # Reset file pointer to beginning\n",
        "        next(csv.reader(file)) # Skip header again\n",
        "        for row in csv.reader(file):\n",
        "            for i, item in enumerate(row):\n",
        "                print(f\"{item:<{column_widths[i]}}\", end=\"  \")\n",
        "            print()\n",
        "\n",
        "# Example usage\n",
        "file_path = os.path.join(jobname, f'boltz_results_{jobname}', 'predictions', jobname, 'stats.csv')\n",
        "print_csv_table(file_path)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "jpTE50xOjQ5R"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Save your data!\n",
        "This last step produces a compressed file called a .zip file containing 4 files:\n",
        "\n",
        "1. the structural model in .cif format, the confidence of the relative placement of each amino acid can be see if the protein is colored by b-factor.\n",
        "2. a parameters file in .yaml format\n",
        "3. an image showing the number of sequences for each protein and the identity of the sequence\n",
        "4.  an .a3m file which contains the sequences used in modeling"
      ],
      "metadata": {
        "id": "G2d5EckbMqNS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Download results\n",
        "# Import necessary modules\n",
        "import os\n",
        "import zipfile\n",
        "from google.colab import files\n",
        "import glob\n",
        "\n",
        "# Ensure 'jobname' variable is defined\n",
        "# jobname = 'test_abcde'  # Uncomment and set if not already defined\n",
        "\n",
        "# Name of the zip file\n",
        "zip_filename = f\"results_{jobname}.zip\"\n",
        "\n",
        "# Create a zip file and add the specified files without preserving directory structure\n",
        "with zipfile.ZipFile(zip_filename, 'w') as zipf:\n",
        "    coverage_png_files = glob.glob(os.path.join(jobname, '*_coverage.png'))\n",
        "    a3m_files = glob.glob(os.path.join(jobname, '*.a3m'))\n",
        "    for file in coverage_png_files + a3m_files:\n",
        "        arcname = os.path.basename(file)  # Use only the file name\n",
        "        zipf.write(file, arcname=arcname)\n",
        "\n",
        "    cif_files = glob.glob(os.path.join(jobname, f'boltz_results_{jobname}', 'predictions', jobname, '*.cif'))\n",
        "    for file in cif_files:\n",
        "        arcname = os.path.basename(file)  # Use only the file name\n",
        "        zipf.write(file, arcname=arcname)\n",
        "\n",
        "    hparams_file = os.path.join(jobname, f'boltz_results_{jobname}', 'lightning_logs', 'version_0', 'hparams.yaml')\n",
        "    if os.path.exists(hparams_file):\n",
        "        arcname = os.path.basename(hparams_file)  # Use only the file name\n",
        "        zipf.write(hparams_file, arcname=arcname)\n",
        "    else:\n",
        "        print(f\"Warning: {hparams_file} not found.\")\n",
        "\n",
        "    json_files = glob.glob(os.path.join(jobname, f'boltz_results_{jobname}', 'predictions', jobname, '*.json'))\n",
        "    for file in json_files:\n",
        "        arcname = os.path.basename(file)  # Use only the file name\n",
        "        zipf.write(file, arcname=arcname)\n",
        "\n",
        "    csv_files = glob.glob(os.path.join(jobname, f'boltz_results_{jobname}', 'predictions', jobname, '*.csv'))\n",
        "    for file in csv_files:\n",
        "        arcname = os.path.basename(file)  # Use only the file name\n",
        "        zipf.write(file, arcname=arcname)\n",
        "\n",
        "# Download the zip file\n",
        "files.download(zip_filename)\n"
      ],
      "metadata": {
        "id": "jdSBSTOpaULF",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Next steps\n",
        "Much like AlphaFold3 models the .cif can be viewed in visualizing tools such a PyMOL or Mol*. These models are great for generating hypotheses to test in wet-lab experiments or for further modeling.\n",
        "\n",
        "A few words of caution, modeling mutants and comparing the models to the prevalent sequence is difficult to do with modeling since the structure always pushes toward good quality protein structures. If you want to see the effects of a mutation computationally, you are encouraged to do molecular dynamics or similar techniques which move the atoms and look at relative stability. [One way to do MD](https://colab.research.google.com/drive/1-Nr3oze1IWND4-eTci9TyykFhtx9fVuX?usp=sharing)."
      ],
      "metadata": {
        "id": "pY6dfgoeq8vZ"
      }
    }
  ]
}